crates/env/src/engine/e2e/xts.rs

```diff
@@ -0,0 +1,343 @@
+#[derive(Debug, scale::Encode, scale::Decode)]
+pub struct Call<C: subxt::Config, B> {
+    dest: ::subxt::ext::sp_runtime::MultiAddress<C::AccountId, ()>,
+    #[codec(compact)]
+    value: B,
+    #[codec(compact)]
+    gas_limit: crate::types::Gas,
+    storage_deposit_limit: Option<B>,
+    data: Vec<u8>,
+}
+
+/// A struct that encodes RPC parameters required to instantiate a new smart contract.
+#[derive(serde::Serialize, scale::Encode)]
+#[serde(rename_all = "camelCase")]
+struct InstantiateRequest<C: subxt::Config, E: Environment> {
+    origin: C::AccountId,
+    value: E::Balance,
+    gas_limit: crate::types::Gas,
+    storage_deposit_limit: Option<E::Balance>,
+    code: Code,
+    data: Vec<u8>,
+    salt: Vec<u8>,
+}
+
+/// Reference to an existing code hash or a new Wasm module.
+#[derive(serde::Serialize, scale::Encode)]
+#[serde(rename_all = "camelCase")]
+enum Code {
+    /// A Wasm module as raw bytes.
+    Upload(Vec<u8>),
+    #[allow(unused)]
+    /// The code hash of an on-chain Wasm blob.
+    Existing(H256),
+}
+
+/// A struct that encodes RPC parameters required for a call to a smart contract.
+///
+/// Copied from [`pallet-contracts-rpc`].
+#[derive(serde::Serialize, scale::Encode)]
+#[serde(rename_all = "camelCase")]
+struct RpcCallRequest<C: subxt::Config, E: Environment> {
+    origin: C::AccountId,
+    dest: C::AccountId,
+    value: E::Balance,
+    gas_limit: crate::types::Gas,
+    storage_deposit_limit: Option<E::Balance>,
+    input_data: Vec<u8>,
+}
+
+/// Provides functions for interacting with the `pallet-contracts` API.
+pub struct ContractsApi<C: subxt::Config, E: Environment> {
+    pub client: OnlineClient<C>,
+    ws_client: WsClient,
+    _phantom: PhantomData<fn() -> (C, E)>,
+}
+
+impl<C, E> ContractsApi<C, E>
+where
+    C: subxt::Config,
+    C::AccountId: Into<C::Address> + serde::de::DeserializeOwned,
+    <C::ExtrinsicParams as ExtrinsicParams<C::Index, C::Hash>>::OtherParams: Default,
+
+    C::Signature: From<sr25519::Signature>,
+    <C::Signature as Verify>::Signer: From<sr25519::Public>,
+    <C::Signature as Verify>::Signer:
+        From<sr25519::Public> + IdentifyAccount<AccountId = C::AccountId>,
+    sr25519::Signature: Into<C::Signature>,
+
+    E: Environment,
+    E::Balance: scale::Encode,
+
+    Call<C, E::Balance>: scale::Encode,
+    InstantiateWithCode<E::Balance>: scale::Encode,
+{
+    /// Creates a new [`ContractsApi`] instance.
+    pub async fn new(client: OnlineClient<C>, url: &str) -> Self {
+        let ws_client =
+            WsClientBuilder::default()
+                .build(&url)
+                .await
+                .unwrap_or_else(|err| {
+                    panic!("error on ws request: {:?}", err);
+                });
+
+        Self {
+            client,
+            ws_client,
+            _phantom: Default::default(),
+        }
+    }
+
+    /// Dry runs the instantiation of the given `code`.
+    pub async fn instantiate_with_code_dry_run(
```

Comment:
This looks to me very similar to the same function definition in `cargo-contract`. Wouldn't it be good to import this method from `cargo-contract` for the sake of single point of truth and DRY?

Reply:
Thanks for flagging it. My view is that in most cases DRY makes total sense, but as you already noted we would couple `cargo-contract` and ink! quite closely together by introducing a dependency here. I would rather have clear boundaries between the two for now.